### PR TITLE
fix: fails to install from private registry (Azure Devops)

### LINKED ISSUE #1815: Fails to install from private registry (Azure Devops)
### pnpm version:
3.2.0

### Code to reproduce the issue:
.npmrc:
```
@{scope}:registry=https://{organisation}.pkgs.visualstudio.com/_packaging/{project}/npm/registry/
//{organisation}.pkgs.visualstudio.com/_packaging/{project}/npm/registry/:username=packageWriteAccess
//{organisation}.pkgs.visualstudio.com/_packaging/{project}/npm/registry/:_password=***
//{organisation}.pkgs.visualstudio.com/_packaging/{project}/npm/registry/:always-auth=true
```

Run:
`pnpm install`

### Expected behavior:
Dependencies - including from private registries - are installed.

### Actual behavior:
Downloading *.tgz file from private registry fails:
` ERROR  400 Authentication information is not given in the correct format. Check the value of Authorization header.: https://{organisation}.pkgs.visualstudio.com/_packaging/{project}/npm/registry/@{scope}/{package}/-/{package}-4.0.12497.tgz`

### Additional information:
- `node -v` prints: `v11.10.0`
- Windows
- Works with pnpm version 2.17.8
- Works with `npm install`

### PR DESCRIPTION
**Pull request by zkochan**
Authorization headers should be removed on redirects.

close #1815

TODO:
- [x] add a test

cc @aquark, @schummar, @christiango

### CODE REVIEW

File: packages/fetch-from-npm-registry/src/index.ts [hunk 1]
```diff
@@ -54,14 +54,27 @@ export default function (
...
+    while (true) {
+      let response = await fetch(url, {
+        agent,
+        // if verifying integrity, node-fetch must not decompress
+        compress: false,
+        headers,
+        redirect: 'manual',
+        retry: defaultOpts.retry,
+      })
+      if (response.status !== 302 || redirects >= 20) return response
```
Comment (octogonz) [inline packages/fetch-from-npm-registry/src/index.ts]: Should 20 be a constant?

File: packages/fetch-from-npm-registry/src/index.ts [hunk 2]
```diff
@@ -54,14 +54,27 @@ export default function (
...
+
+      // This is a workaround to remove authorization headers on redirect.
+      // It is needed until node-fetch doesn't fix this
+      // or support a way to do it via an option.
```
Comment (octogonz) [inline packages/fetch-from-npm-registry/src/index.ts]: Did you mean to say "until node-fetch fixes this"?

File: packages/fetch-from-npm-registry/src/index.ts (hunk 1 above)
Comment (aquark) [inline packages/fetch-from-npm-registry/src/index.ts]: To fix my issue, this needs to also handle 307 redirects. Ideally it should handle all 3xx redirects. It turns out there is a `fetch.isRedirect` method which handles this. The following works for me:

```suggestion
      if (!fetch.isRedirect(response.status) || redirects >= 20) return response
```

**Comment by aquark** [comment]
Confirmed the latest commit fixes the issue for me.

**Comment by zkochan** [comment]
:ship: [pnpm@3.3.2](https://github.com/pnpm/pnpm/releases/tag/v3.3.2)

**Comment by christiango** [comment]
Can confirm things are working nicely for me. Thanks for the quick turnaround!!

**Pull request merged**